consensus/src/epoch_manager.rs

```diff
@@ -263,9 +251,48 @@ impl EpochManager {
                     vec![1; proposers.len()]
                 };
 
+                // First two epochs are immediatelly skipped after a few transactions, so we don't want to look at their history.
+                let epoch_to_proposers = if use_history_from_previous_epoch && epoch_state.epoch > 2
+                {
+                    self.storage
+                        .aptos_db()
+                        .get_epoch_ending_ledger_infos(epoch_state.epoch - 2, epoch_state.epoch)
+                        .and_then(|proof| {
+                            ensure!(proof.ledger_info_with_sigs.len() == 2);
+                            let prev_epoch_state = proof.ledger_info_with_sigs[0]
+                                .ledger_info()
+                                .next_epoch_state()
+                                .ok_or_else(|| anyhow::anyhow!("no prev_epoch_state"))?;
+                            let cur_epoch_state = proof.ledger_info_with_sigs[1]
+                                .ledger_info()
+                                .next_epoch_state()
+                                .ok_or_else(|| anyhow::anyhow!("no cur_epoch_state"))?;
+                            ensure!(cur_epoch_state.epoch == epoch_state.epoch, "fetched epoch_ending ledger_infos are for a wrong epoch {} vs {}", cur_epoch_state.epoch, epoch_state.epoch);
+                            let prev_epoch_proposers = prev_epoch_state
+                                .verifier
+                                .get_ordered_account_addresses_iter()
+                                .collect::<Vec<_>>();
+                            let cur_epoch_proposers = cur_epoch_state
+                                .verifier
+                                .get_ordered_account_addresses_iter()
+                                .collect::<Vec<_>>();
+                            ensure!(proposers == cur_epoch_proposers, "proposers from state and fetched epoch_ending ledger_infos are missaligned");
+                            Ok(HashMap::from([
+                                (prev_epoch_state.epoch, prev_epoch_proposers),
+                                (cur_epoch_state.epoch, cur_epoch_proposers),
+                            ]))
+                        })
+                        .unwrap_or_else(|err| {
+                            error!("Couldn't create leader reputation with history across epochs, {:?}", err);
+                            HashMap::from([(epoch_state.epoch, proposers)])
+                        })
+                } else {
+                    HashMap::from([(epoch_state.epoch, proposers)])
```

Comment:
maybe it's just my preference, I find it may be more readable like this:
```
  let mut epoch_to_proposers = HashMap::from([(epoch_state.epoch, proposers)]);
  if use_history_from_previous_epoch {
    if let Ok(history_epoch_to_proposers) = self.storage.aptos_db().get_epoch_ending_ledger_infos().and_then() {
      epoch_to_proposers = history_epoch_to_proposers;
    }
  }
```